### PR TITLE
Add VirtualMachineOptions to KV configuration

### PR DESCRIPTION
**Pull request by fossedihelm**
**What this PR does / why we need it**:
With https://github.com/kubevirt/kubevirt/pull/9572  freePageReporting will be enabled
for only those vmis which does not request any high performance features.

This PR allows cluster admins to define the default behavior for for the virtual machines.
Currently, the new config will only allow cluster admins to opt-out
from using freePageReporting for all the vmis in the cluster, but it can be used for 
future virtual machine behavior to be defined at cluster level.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Introduce cluster configuration `VirtualMachineOptions` to specify virtual machine behavior at cluster level
```

### CODE REVIEW

File: staging/src/kubevirt.io/api/core/v1/types.go [hunk 1]
```diff
@@ -2397,6 +2398,30 @@ type SeccompConfiguration struct {
 	VirtualMachineInstanceProfile *VirtualMachineInstanceProfile `json:"virtualMachineInstanceProfile,omitempty"`
 }
 
+// WorkloadMemoryOverCommitmentProfile holds the workload memory over-commitment profile for Kubevirt.
+type WorkloadMemoryOverCommitmentProfile struct {
```
Comment (acardace) [inline staging/src/kubevirt.io/api/core/v1/types.go:2402]: While I like what you have done here (a struct for future expansion) I don't think the name is spot on as FPR is not technically memory over-commitment.

Maybe something like this is more appropriate?
```
type WorkloadMemoryProfile struct {
    Strategy *Strategy
    Custom *WorkloadMemoryConfiguration
}

type WorkloadMemoryConfiguration struct {
	FreePageReporting *bool `json:"freePageReporting,omitempty"`
}

```

**Comment by fossedihelm** [comment]
@vladikr @acardace I adjusted the PR following the outcome from the offline discussion. Thanks

**Review by xpivarc — COMMENTED** [review]
I would be interested in the background on why would one opt out of this.

File: pkg/handler-launcher-com/cmd/v1/cmd.proto [hunk 2]
```diff
@@ -94,7 +99,7 @@ message VirtualMachineOptions {
   repeated string PreallocatedVolumes = 3;
   Topology topology = 4;
   map<string, DiskInfo> DisksInfo = 5;
-  bool ExpandDisksEnabled = 6;
+  ClusterConfig clusterConfig = 6;
```
[the remark this answers is not in the document]
Comment (fossedihelm) [reply pkg/handler-launcher-com/cmd/v1/cmd.proto]: Thanks! Restored

**Comment by fossedihelm** [comment]
> I would be interested in the background on why would one opt out of this.

I would be inclined to say in those cases where CPU overhead is more important than the memory one. WDYT? Thanks

**Comment by vladikr** [comment]
/approve 
Looks good to me, in general.

/hold
@xpivarc @acardace what do you think?

File: staging/src/kubevirt.io/api/core/v1/types.go [hunk 3]
```diff
@@ -2400,6 +2401,17 @@ type SeccompConfiguration struct {
 	VirtualMachineInstanceProfile *VirtualMachineInstanceProfile `json:"virtualMachineInstanceProfile,omitempty"`
 }
 
+// VirtualMachineOptions holds the cluster level information regarding the virtual machine.
+type VirtualMachineOptions struct {
+	// DisableFreePageReporting disable the free page reporting of
+	// memory balloon device https://libvirt.org/formatdomain.html#memory-balloon-device.
+	// This will have effect only if AutoattachMemBalloon is not false and the vmi is not
+	// requesting any high performance feature, in which free page reporting is always disabled.
+	DisableFreePageReporting *DisableFreePageReporting `json:"disableFreePageReporting,omitempty"`
```
Comment (xpivarc) [inline staging/src/kubevirt.io/api/core/v1/types.go:2410]: Nit: I would describe what `high performance feature` is.

File: pkg/virt-handler/options.go [hunk 4]
```diff
@@ -32,6 +33,15 @@ func virtualMachineOptions(
 			Version:      smbios.Version,
 		}
 	}
+
+	if clusterConfig != nil {
+		options.ExpandDisksEnabled = clusterConfig.ExpandDisksEnabled()
+		options.ClusterConfig = &cmdv1.ClusterConfig{
+			ExpandDisksEnabled:        clusterConfig.ExpandDisksEnabled(),
+			FreePageReportingDisabled: clusterConfig.GetVirtualMachineOptions() != nil && clusterConfig.GetVirtualMachineOptions().DisableFreePageReporting != nil,
```
Comment (xpivarc) [inline pkg/virt-handler/options.go:41]: How do we want to interpret `DisableFreePageReporting` == false?

**Review by xpivarc — COMMENTED** [review]
Looks good. I left just few minor comments

File: staging/src/kubevirt.io/api/core/v1/types.go (hunk 3 above)
Comment (fossedihelm) [reply staging/src/kubevirt.io/api/core/v1/types.go:2410]: Thanks. I'll add it

File: pkg/virt-config/configuration_test.go [hunk 5]
```diff
@@ -243,6 +243,32 @@ var _ = Describe("test configuration", func() {
 		Entry("when empty, GetEmulatedMachines should return the defaults with ppc64le", "ppc64le", []string{}, []string{}, []string{}, strings.Split(virtconfig.DefaultPPC64LEEmulatedMachines, ",")),
 	)
 
+	DescribeTable("when virtualMachineOptions", func(virtualMachineOptions *v1.VirtualMachineOptions) {
```
[the remark this answers is not in the document]
Comment (fossedihelm) [reply pkg/virt-config/configuration_test.go:246]: What do you mean exactly?

**Comment by fossedihelm** [comment]
@xpivarc Comments addressed. PTAL. Thanks

**Review by xpivarc — APPROVED** [review]

**Pull request merged**